### PR TITLE
Update froidure-pin-base-hpp for v1/3

### PR DESCRIPTION
**Pull request by james-d-mitchell**

### CODE REVIEW

File: src/froidure-pin-base.cpp [hunk 1]
```diff
@@ -0,0 +1,725 @@
...
+    ...        [0,  0,  1,  0],
+    ...        [0,  0,  1,  0]]),
+    ... BMat8([[0,  0,  0,  1],
+    ...        [0,  0,  0,  1],
+    ...        [0,  0,  0,  1],
+    ...        [0,  0,  0,  1]]))
+    >>> S.size()
+    4
+    >>> list(S.rules())
+)pbdoc");
```
Comment (Joseph-Edwards) [inline src/froidure-pin-base.cpp:161]: Should there be output here?

File: src/froidure-pin-base.cpp [hunk 2]
```diff
@@ -0,0 +1,725 @@
...
+
+    thing.def("right_cayley_graph",
+              &FroidurePinBase::right_cayley_graph,
+              R"pbdoc(
+Returns the fully enumerated right Cayley graph.
+
+:complexity:
+  At worst :math:`O(|S|n)` where :math:`S` is the semigroup represented
+  by ``self``, and :math:`n` is the return value of
+  :any:`FroidurePin.number_of_generators`.
```
Comment (Joseph-Edwards) [inline src/froidure-pin-base.cpp:679]: This should be below `:rtype:`

File: src/froidure-pin-base.cpp [hunk 3]
```diff
@@ -0,0 +1,725 @@
...
+        R"pbdoc(
+Returns an iterator containing the so-far enumerated normal forms (if any).
+This function returns an iterator pointing at the normal form of the first
+element of the semigroup represented by a :any:`FroidurePinBase` instance (if
+any). This function does not perform any enumeration of the :any:`FroidurePin`.
+If you want to obtain the complete set of rules, then use :any:`normal_forms`
+instead.
+
+:returns:
+  An iterator yielding a :any:`List[int]`.
```
Comment (Joseph-Edwards) [inline src/froidure-pin-base.cpp:110]:
```suggestion
  An iterator yielding a ``List[int]``.
```

File: src/froidure-pin-base.cpp [hunk 4]
```diff
@@ -0,0 +1,725 @@
...
+Returns an iterator yielding the so-far enumerated rules.
+
+Returns an iterator yielding the rules in a confluent
+terminating rewriting system defining a semigroup isomorphic to the one
+defined by ``self``. This function performs a full enumeration of the
+:any:`FroidurePin` object. If you want to obtain the current set of rules
+without triggering any enumeration, then use :any:`current_rules` instead.
+
+:returns:
+    An iterator yielding :any:`Tuple[List[int],List[int]]` .
```
Comment (Joseph-Edwards) [inline src/froidure-pin-base.cpp:201]:
```suggestion
    An iterator yielding ``Tuple[List[int],List[int]]`` .
```

File: src/froidure-pin-base.cpp [hunk 5]
```diff
@@ -0,0 +1,725 @@
...
+namespace libsemigroups {
+
+  void init_froidure_pin_base(py::module& m) {
+    py::class_<FroidurePinBase> thing(m,
+                                      "FroidurePinBase",
+                                      R"pbdoc(
+Base class for :any:`FroidurePin` containing non-element specific data and
+member functions.
+
+.. seealso::  :any:`FroidurePin`.)pbdoc");
```
Comment (Joseph-Edwards) [inline src/froidure-pin-base.cpp:49]: Is it worth mentioning here that it's an abstract base class, or at least saying you can't construct a `FroidurePinBase`?

**Review by Joseph-Edwards — CHANGES_REQUESTED** [review]

File: src/froidure-pin-base.cpp (hunk 1 above)
Comment (james-d-mitchell) [reply src/froidure-pin-base.cpp:161]: `FroidurePinBase` is an abstract class, and I haven't yet completed `FroidurePin` so the output is pending :)

File: src/froidure-pin-base.cpp (hunk 5 above)
Comment (james-d-mitchell) [reply src/froidure-pin-base.cpp:49]: Yeah almost certainly. Will do later in the afternoon, thanks for the review! @Joseph-Edwards

**Comment by james-d-mitchell** [comment]
I've made the changes you suggested, should be good to go now.

**Review by Joseph-Edwards — APPROVED** [review]

**Pull request merged**